overlord/snapstate/autorefresh.go

```diff
@@ -974,3 +974,127 @@ func MockRefreshCandidate(snapSetup *SnapSetup) interface{} {
 		SnapSetup: *snapSetup,
 	}
 }
+
+func incrementSnapRefreshFailures(st *state.State, snapsup *SnapSetup) error {
+	var snapst SnapState
+	err := Get(st, snapsup.InstanceName(), &snapst)
+	if err != nil {
+		return err
+	}
+
+	// Update refresh failure information for snap revision
+	if snapst.RefreshFailures == nil {
+		// Initialize RefreshFailures
+		snapst.RefreshFailures = &snap.RefreshFailuresInfo{}
+	}
+	if snapst.RefreshFailures.ToRevision == snapsup.Revision() {
+		snapst.RefreshFailures.FailureCount++
+	} else {
+		snapst.RefreshFailures.ToRevision = snapsup.Revision()
+		snapst.RefreshFailures.FailureCount = 1
+	}
+	snapst.RefreshFailures.LastFailureTime = timeNow()
+	Set(st, snapsup.InstanceName(), &snapst)
+
+	delay := computeSnapRefreshDelay(snapst.RefreshFailures)
+	logger.Noticef("snap %q auto-refresh to revision %s has failed, next auto-refresh attempt will be in %v hours", snapsup.InstanceName(), snapsup.Revision(), delay.Hours())
+	return nil
+}
+
+func processFailedAutoRefresh(chg *state.Change, _ state.Status, new state.Status) {
+	if chg.Kind() != "auto-refresh" || new != state.ErrorStatus {
+		return
+	}
+
+	var failedSnapNames []string
+	for _, t := range chg.Tasks() {
+		// We only care about snaps that failed after unlink-current-snap because
+		// this indicates (with high probability) that something related to the snap
+		// itself is broken.
+		if t.Kind() != "unlink-current-snap" || t.Status() != state.UndoneStatus {
```

Comment:
This is a smart way of doing this. Nice!

Reply:
Thanks!